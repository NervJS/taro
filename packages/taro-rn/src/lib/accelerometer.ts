import { Accelerometer } from 'expo-sensors'

const accCase: any = {
  callbacks: [],
}
const intervalMap: any = {
  game: 20,
  ui: 60,
  normal: 200
}

function offAccelerometerChange(fnc?: Taro.onAccelerometerChange.Callback): void {
  if (fnc && typeof fnc === 'function') {
    accCase.callbacks = accCase.callbacks.filter((cb: (...args: any[]) => any) => cb !== fnc)
  } else if (fnc === undefined) {
    accCase.callbacks = []
  } else {
    console.warn('offAccelerometerChange failed')
  }
}

function onAccelerometerChange(fnc: Taro.onAccelerometerChange.Callback): void {
  accCase.callbacks.push(fnc)
}

/**
 * 开始监听加速度数据。
 * @param {Object} opts
 * @param {string} [opts.interval='normal'] 监听加速度数据回调函数的执行频率
 */
function startAccelerometer(opts: Taro.startAccelerometer.Option = {}): Promise<Taro.General.CallbackResult> {
  const { interval = 'normal', success, fail, complete } = opts
  accCase.interval = interval
  const res = { errMsg: 'startAccelerometer:ok' }
  try {
    // 适配微信小程序行为：重复 start 失败
    if (accCase.listener) {
      throw new Error('startAccelerometer:fail')
    }
    accCase.listener = Accelerometer.addListener((e: Taro.onAccelerometerChange.Result) => {
      accCase.callbacks.forEach((cb: (...args: any[]) => any) => {
        cb?.(e)
      });
    })
    success?.(res)
    complete?.(res)
  } catch (error) {
    res.errMsg = 'startAccelerometer:fail'
    fail?.(res)
    complete?.(res)
    return Promise.reject(res)
  }
  Accelerometer.setUpdateInterval(intervalMap[interval])
  return Promise.resolve(res)
}

/**
 * 停止监听加速度数据
 * @param opts
 */
function stopAccelerometer(opts: Taro.stopAccelerometer.Option = {}): Promise<Taro.General.CallbackResult> {
  const { success, fail, complete } = opts
  const res = { errMsg: 'stopAccelerometer:ok' }
  try {
    accCase.listener.remove()
    accCase.listener = null
    success?.(res)
    complete?.(res)
    return Promise.resolve(res)
  } catch (error) {
    res.errMsg = 'stopAccelerometer:fail'
    fail?.(res)
    complete?.(res)
    return Promise.reject(res)
  }
}

export {
  onAccelerometerChange,
  offAccelerometerChange,
  startAccelerometer,
  stopAccelerometer
}
