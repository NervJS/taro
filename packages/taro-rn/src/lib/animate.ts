import { Animated, Easing } from "react-native";

export const animate = (selector, keyframes, duration, callback) => {
  console.log('hello animate')
  // console.log(selector['_internalFiberInstanceHandleDEV'].memoizedProps.style)
  // console.log(keyframes)
  // keyframes.forEach(item => {
  //   selector['_internalFiberInstanceHandleDEV'].memoizedProps.style = item
  // })
  // console.log(selector)
  // return {}
}
