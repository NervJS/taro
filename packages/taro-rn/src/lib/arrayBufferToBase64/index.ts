import { fromByteArray } from 'base64-js'

export function arrayBufferToBase64(arrayBuffer: ArrayBuffer): string {
  const buffer = new Uint8Array(arrayBuffer)
  return fromByteArray(buffer)
}