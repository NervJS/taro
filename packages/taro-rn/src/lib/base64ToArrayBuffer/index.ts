import { toByteArray } from 'base64-js'

export function base64ToArrayBuffer(base64: string): ArrayBuffer {
  return toByteArray(base64)
}