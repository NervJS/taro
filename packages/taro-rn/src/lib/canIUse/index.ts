export function canIUse(): boolean {
  return true
}
