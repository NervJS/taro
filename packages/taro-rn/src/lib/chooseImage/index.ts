import SYImagePicker from 'react-native-syan-image-picker'

import { shouleBeObject } from '../../utils'
import { showActionSheet } from '../showActionSheet'

const res = { errMsg: 'chooseImage:ok' }
const defaultOptions = {
  count: 9,
  sizeType: ['original', 'compressed'],
  sourceType: ['album', 'camera']
}

function getRes(images) {
  const tempFilePaths: string[] = []
  const tempFiles: any[] = []
  images.forEach((item: any) => {
    tempFilePaths.push(item.uri)
    tempFiles.push({
      path: item.uri,
      size: item.size
    })
  })
  return { tempFilePaths, tempFiles }
}

function openCamera(options: Taro.chooseImage.Option): Promise<Taro.chooseImage.SuccessCallbackResult> {
  const { success, fail, complete } = options
  return new Promise((resolve, reject) => {
    SYImagePicker.openCamera({}, (err, photo: any) => {
      if (err) {
        const res = {
          errMsg: err
        }
        fail?.(res)
        complete?.(res)
        reject(res)
      } else {
        try {
          const data = Object.assign({}, getRes(photo), res)
          success?.(data)
          complete?.(data)
          resolve(data)
        } catch (error) {
          // 错误
          reject(error)
        }
      }
    })
  })
}

function openPicker(options: Taro.chooseImage.Option): Promise<Taro.chooseImage.SuccessCallbackResult> {
  const { count: imageCount, success, complete, fail } = options
  return new Promise((resolve, reject) => {
    SYImagePicker.showImagePicker({
      imageCount,
    }, (err, photos: any) => {
      if (err) {
        const res = {
          errMsg: err
        }
        fail?.(res)
        complete?.(res)
        reject(res)
      } else {
        try {
          const data = Object.assign({}, getRes(photos), res)
          success?.(data)
          complete?.(data)
          resolve(data)
        } catch (error) {
          reject(error)
        }
      }
    })
  })
}

/**
 * @description Display the image picker.
 * @param options
 * @returns {Promise}
 */
function showImagePicker(options: Taro.chooseImage.Option):Promise<Taro.chooseImage.SuccessCallbackResult> {
  const { fail, complete } = options
  return new Promise((resolve, reject) => {
    showActionSheet({
      itemList: ['拍照', '从手机相册选择'],
      success(res) {
        console.log(res.tapIndex)
        if (res.tapIndex === 0) {
          // 不添加 setTimeout 会有一定概率卡死，肯能是与 sibling.destroy() 冲突
          setTimeout(() => {
            openCamera(options).then(res => {
              resolve(res)
            }).catch(e => {
              fail?.({ errMsg: `chooseImage:${e.message}` })
              complete?.({ errMsg: `chooseImage:${e.message}` })
              reject(new Error(`chooseImage:${e.message}`))
            })
          }, 100)
        }
        if (res.tapIndex === 1) {
          setTimeout(() => {
            openPicker(options).then(res => {
              resolve(res)
            }).catch(e => {
              fail?.({ errMsg: `chooseImage:${e.message}` })
              complete?.({ errMsg: `chooseImage:${e.message}` })
              reject(new Error(`chooseImage:${e.message}`))
            })
          }, 100)
        }
      },
      fail(res) {
        console.log('error: ', res.errMsg)
      }
    })
  })
}

/**
 * @deprecated
 * @description 从本地相册选择图片或使用相机拍照。
 *  sizeType ❌
 * @param options
 * @returns {Promise<*>}
 */
export function chooseImage(options: Taro.chooseImage.Option): Promise<Taro.chooseImage.SuccessCallbackResult> {
  // options must be an Object
  const isObject = shouleBeObject(options)
  if (!isObject.res) {
    const res = { errMsg: `chooseImage${isObject.msg}` }
    console.error(res.errMsg)
    return Promise.reject(res)
  }

  options = Object.assign({}, defaultOptions, options)
  const { sourceType } = options

  switch (JSON.stringify(sourceType)) {
    case '[\'camera\']':
    case '["camera"]':
      return openCamera(options)
    case '[\'album\']':
    case '["album"]':
      return openPicker(options)
    default:
      return showImagePicker(options)
  }
}
