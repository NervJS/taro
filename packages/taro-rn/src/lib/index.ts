// 由 getLibList.js 脚本生成, 不要进行手动修改, 请不要手动修改
export * from './ENV_TYPE'
export * from './arrayBufferToBase64'
export * from './authorize'
export * from './base64ToArrayBuffer'
export * from './canIUse'
export * from './chooseImage'
export * from './chooseVideo'
export * from './clearStorage'
export * from './compressImage'
export * from './connectSocket'
export * from './createCameraContext'
export * from './createInnerAudioContext'
export * from './createVideoContext'
export * from './downloadFile'
export * from './getClipboardData'
export * from './getEnv'
export * from './getFileInfo'
export * from './getImageInfo'
export * from './getLocation'
export * from './getNetworkType'
export * from './getRecorderManager'
export * from './getSavedFileInfo'
export * from './getSavedFileList'
export * from './getScreenBrightness'
export * from './getSetting'
export * from './getStorage'
export * from './getStorageInfo'
export * from './getSystemInfo'
export * from './getSystemInfoSync'
export * from './hideKeyboard'
export * from './hideLoading'
export * from './hideToast'
export * from './makePhoneCall'
export * from './offAccelerometerChange'
export * from './offGyroscopeChange'
export * from './offKeyboardHeightChange'
export * from './offNetworkStatusChange'
export * from './offWindowResize'
export * from './onAccelerometerChange'
export * from './onDeviceMotionChange'
export * from './onGyroscopeChange'
export * from './onKeyboardHeightChange'
export * from './onNetworkStatusChange'
export * from './onWindowResize'
export * from './openSetting'
export * from './openUrl'
export * from './previewImage'
export * from './removeSavedFile'
export * from './removeStorage'
export * from './request'
export * from './saveFile'
export * from './saveImageToPhotosAlbum'
export * from './saveVideoToPhotosAlbum'
export * from './scanCode'
export * from './setClipboardData'
export * from './setKeepScreenOn'
export * from './setScreenBrightness'
export * from './setStorage'
export * from './showActionSheet'
export * from './showLoading'
export * from './showModal'
export * from './showToast'
export * from './startAccelerometer'
export * from './startDeviceMotionListening'
export * from './startGyroscope'
export * from './stopAccelerometer'
export * from './stopDeviceMotionListening'
export * from './stopGyroscope'
export * from './uploadFile'
export * from './vibrateLong'
export * from './vibrateShort'
