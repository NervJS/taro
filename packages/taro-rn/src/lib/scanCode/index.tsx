import { SafeAreaView } from 'react-native-safe-area-context'
import { Camera } from 'expo-camera'
import { BarCodeScanner } from 'expo-barcode-scanner'
import { BackHandler, Image, TouchableOpacity, StyleSheet, View, Dimensions, Platform } from 'react-native'
import iconClose from './icon_close.png'
import iconPic from './icon_pic.png'
import * as Permissions from 'expo-permissions'
import RootSiblings from 'react-native-root-siblings'
import { chooseMedia, MEDIA_TYPE } from '../media'
import React from 'react'

export let scannerView

const codeMap = {
  aztec: 'AZTEC',
  codabar: 'CODABAR',
  code39: 'CODE_39',
  code93: 'CODE_93',
  code128: 'CODE_128',
  code39mod43: 'CODE_93', // 未准确对应
  datamatrix: 'DATA_MATRIX',
  ean13: 'EAN_13',
  ean8: 'EAN_8',
  interleaved2of5: 'ITF', // 未准确对应
  itf14: 'ITF', // 未准确对应
  maxicode: 'MAXICODE',
  pdf417: 'PDF_417',
  rss14: 'RSS_14',
  rssexpanded: 'RSS_EXPANDED',
  upc_a: 'UPC_A',
  upc_e: 'UPC_E',
  upc_ean: 'UPC_EAN_EXTENSION',
  qr: 'QR_CODE'
}

const typeMap = {
  barCode: ['aztec', 'codabar', 'code39', 'code93', 'code128', 'code39mod43', 'ean13', 'ean8', 'interleaved2of5', 'itf14', 'maxicode', 'rss14', 'rssexpanded', 'upc_a', 'upc_e', 'upc_ean'],
  qrCode: ['qr'],
  datamatrix: ['datamatrix'],
  pdf417: ['pdf417']
}

const BarCodeType = BarCodeScanner.Constants.BarCodeType

function findKey (value:string, data, compare = (a, b) => a === b):string {
  return Object.keys(data).find(k => compare(data[k], value)) || ''
}

const { width, height } = Dimensions.get('window')

function getBarCodeTypes(types:string[]):string[] {
  const result:string[] = []
  for (const t of types) {
    result.push(...typeMap[t].map(type => {
      return BarCodeType[type]
    }))
  }
  return result.filter(r => !!r)
}

function formatCodeType(type:string):keyof Taro.scanCode.QRType {
  return codeMap[findKey(type, BarCodeType)]
}

function safeViewWrapper(element:any) {
  if (Platform.OS === 'ios') {
    return <SafeAreaView>{element}</SafeAreaView>
  }
  return element
}

let backAction: any = null

function hide(view) {
  if (!(view instanceof RootSiblings)) {
    return
  }
  view.destroy()
  BackHandler.removeEventListener('hardwareBackPress', backAction)
}

backAction = () => {
  hide(scannerView)
  return true
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    backgroundColor: '#FFF',
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
    zIndex: 1000
  },
  closeIcon: {
    position: 'absolute',
    left: 20,
    top: 10
  },
  closeImg: {
    width: 25,
    height: 25,
  },
  closeBtnBg: {
    width: 40,
    height: 40,
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center'
  },
  albumIcon: {
    position: 'absolute',
    right: 20,
    bottom: 40
  },
  albumImg: {
    width: 20,
    height: 20,
  },
  albumBtnBg: {
    backgroundColor: 'rgba(0,0,0,0.2)',
    width: 40,
    height: 40,
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center'
  }
})

function scanFromPhoto(callback, errorCallBack) {
  chooseMedia({
    sourceType: ['album'],
    maxDuration: 60,
    camera: 'back',
    success: function (res) {
      const imageUrl = res.tempFilePaths[0]
      if (imageUrl) {
        BarCodeScanner.scanFromURLAsync(imageUrl).then(res => {
          res && res.length > 0 && callback(res[0].data, res[0].type)
        })
      }
    }
  }, MEDIA_TYPE.IMAGES).catch(errorCallBack)
}

export async function scanCode(option: Taro.scanCode.Option = {}): Promise<Taro.scanCode.SuccessCallbackResult> {
  const { success, fail, complete, onlyFromCamera, scanType = ['barCode', 'qrCode'] } = option
  const { status } = await Permissions.askAsync(Permissions.CAMERA)
  if (status !== 'granted') {
    const res = { errMsg: 'Permissions denied!' }
    fail?.(res)
    complete?.(res)
    return Promise.reject(res)
  }
  const barCodeTypes = getBarCodeTypes(scanType)
  return new Promise((resolve, reject) => {
    scannerView = new RootSiblings(
      (<View style={[styles.container]}>
          <Camera
              onBarCodeScanned={({ type, data }: {type: string, data: string}) => {
                const res = {
                  charSet: 'UTF-8', // todo
                  path: '', // todo
                  rawData: '', // todo
                  errMsg: 'scanCode:ok',
                  result: data,
                  scanType: formatCodeType(type),
                }
                success?.(res)
                complete?.(res)
                hide(scannerView)
                resolve(res)
              }}
              barCodeScannerSettings={{
                barCodeTypes,
              }}
              style={{ width, height }}
              ratio='16:9'
          />
          <TouchableOpacity style={styles.closeIcon} onPress={() => hide(scannerView)}>
            {safeViewWrapper(<Image source={iconClose} style={styles.closeImg}/>)}
          </TouchableOpacity>
          {!onlyFromCamera && (<TouchableOpacity style={styles.albumIcon} onPress={() => {
            scanFromPhoto((data, type) => {
              const res = {
                charSet: 'UTF-8', // todo
                path: '', // todo
                rawData: '', // todo
                errMsg: 'scanCode:ok',
                result: data,
                scanType: formatCodeType(type),
              }
              success?.(res)
              complete?.(res)
              hide(scannerView)
              resolve(res)
            }, (err) => {
              const res = {
                errMsg: 'scanCode fail',
                err
              }
              fail?.(res)
              complete?.(res)
              hide(scannerView)
              reject(res)
            })
          }}>
            {safeViewWrapper(<Image source={iconPic} style={styles.albumImg}/>)}
          </TouchableOpacity>)}
        </View>)
    )
    BackHandler.addEventListener('hardwareBackPress', backAction)
  })
}
