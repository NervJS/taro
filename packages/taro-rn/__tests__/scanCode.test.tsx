import { render, waitFor } from '@testing-library/react-native'
import React from 'react'
import { RootSiblingParent } from 'react-native-root-siblings'

import { scanCode } from '../src/lib/scanCode'

describe('scanCode', () => {
  it('should render scanCode success', async () => {
    // @ts-ignore
    const { toJSON, getByLabelText } = render(<RootSiblingParent />)
    scanCode()
    await waitFor(() => getByLabelText('Close'))
    expect(toJSON()).toMatchSnapshot()
  })
})
