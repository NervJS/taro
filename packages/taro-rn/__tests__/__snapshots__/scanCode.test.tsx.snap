// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scanCode should render scanCode success 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "flex": 1,
        "justifyContent": "center",
        "position": "absolute",
        "zIndex": 1000,
      },
    ]
  }
>
  <ViewManagerAdapter_ExpoCamera
    barCodeScannerEnabled={true}
    barCodeScannerSettings={
      {
        "barCodeTypes": [],
      }
    }
    faceDetectorSettings={{}}
    focusDepth={0}
    onBarCodeScanned={[Function]}
    onCameraReady={[Function]}
    onFacesDetected={[Function]}
    onMountError={[Function]}
    onPictureSaved={[Function]}
    onResponsiveOrientationChanged={[Function]}
    style={
      {
        "height": 1334,
        "width": 750,
      }
    }
    zoom={0}
  />
  <View
    accessibilityLabel="Close"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "left": 20,
        "opacity": 1,
        "position": "absolute",
        "top": 10,
      }
    }
  >
    <View
      style={
        {
          "paddingTop": 20,
        }
      }
    >
      <Image
        source={1}
        style={
          {
            "height": 25,
            "marginTop": null,
            "width": 25,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "bottom": 40,
        "opacity": 1,
        "position": "absolute",
        "right": 20,
      }
    }
  >
    <View
      style={
        {
          "paddingTop": 20,
        }
      }
    >
      <Image
        source={1}
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
      />
    </View>
  </View>
</View>
`;
